# Using the PyRockWave tool and the elastic mineral database

One of the features of PyRockWave is that it includes access to a (still incomplete) database of elastic properties of common minerals. This database is not a passive collection of data, but a file of functions written in Python that return the elastic properties of minerals under certain conditions. Although the development of this database is somewhat independent of the PyRockWave tool, it is intended that the various versions of PyRockWave will include the latest version of the database at the time of release. The idea is to provide an up-to-date, well-documented database of elastic properties with rigorous tracking of changes (key to reproducibility).

In this Jupyter notebook, we will show you how to access and use the elastic mineral database through the PyRockWave tool. The steps are as follows:

In [1]:
# import some some Python scientific libraries 
import numpy as np
import pandas as pd
np.set_printoptions(suppress=True)

# import the database
import elastic_database as db

If you then type ``db.`` and press the TAB key, you will get a menu with all the options available in the database. For example, let's extract the properties of alpha quartz at a pressure of 1.8 GPa.

In [2]:
db.alpha_quartz(P=1.8)

<class 'ElasticClass.ElasticProps'>

Mineral Name: Alpha_quartz
Reference Source: https://doi.org/10.1007/s00269-014-0711-z
Crystal System: Trigonal
Pressure (GPa): 1.8
Temperature (°C): 25
Density (g/cm3): 2.763

Elastic Tensor (Cij) in GPa:
[[ 93.52  18.88  24.43  11.     0.     0.  ]
 [ 18.88  93.52  24.43 -11.     0.     0.  ]
 [ 24.43  24.43 129.83   0.     0.     0.  ]
 [ 11.   -11.     0.    62.1    0.     0.  ]
 [  0.     0.     0.     0.    62.1   11.  ]
 [  0.     0.     0.     0.    11.    37.32]]

Tensor decomposition (Browaeys & Chevrot approach):
    Isotropy = 72.4 %
    Anisotropy = 27.6 %

Anisotropy indexes:
    Universal Elastic Anisotropy:           0.558
    Kube's Anisotropy Index (proportional): 0.227

Calculated elastic average properties:
Unit:GPa  Bulk_modulus  Shear_modulus  Poisson_ratio
   Voigt        50.261         48.913          0.133
    Hill        49.551         46.575          0.142
   Reuss        48.842         44.236          0.152

Isotropic sei

As you can see, the command returns a summary of the different properties of alpha quartz at 1.8 GPa pressure, including some average elastic properties, isotropic seismic properties, and several anisotropy measures. If you want to access the function's own documentation you can use the ``help()`` or ``?`` command.

In [3]:
# alternatively you can use help(db.alpha_quartz) or db.alpha_quartz??
db.alpha_quartz?

Signature: db.alpha_quartz(P=1e-05)
Docstring:
Returns the corresponding elastic tensor (GPa) and density
(g/cm3) and other derived elastic properties of alpha quartz
as a function of pressure based on a polynomial fit from
experimental data of Wang et al. (2006) [1]

Caveats
-------
    - C44 elastic term is worse constrained than the others.
    The fitting has an R-squared value of 0.96
    - The function does not account for temperature effects
    and assumes room temperature.

Parameters
----------
P : numeric, optional
    Pressure in GPa. Default value is 1e-5 GPa (RT).

Returns
-------
properties : ElasticProps dataclass
    An object containing the following properties:
    - name: Name of the crystal ('alpha_quartz').
    - crystal_system: Crystal system.
    - temperature: Temperature in degrees Celsius (assumed 25).
    - pressure: Pressure in GPa.
    - density: Density in g/cm3.
    - cijs: 6x6 array representing the elastic tensor.
    - sijs: 6x6 array representing the

You can store this information in a variable and access the different properties individually. For example:

In [4]:
quartz = db.alpha_quartz(1.8)

# show the elastic (stiffness) tensor
quartz.Cij

array([[ 93.52,  18.88,  24.43,  11.  ,   0.  ,   0.  ],
       [ 18.88,  93.52,  24.43, -11.  ,   0.  ,   0.  ],
       [ 24.43,  24.43, 129.83,   0.  ,   0.  ,   0.  ],
       [ 11.  , -11.  ,   0.  ,  62.1 ,   0.  ,   0.  ],
       [  0.  ,   0.  ,   0.  ,   0.  ,  62.1 ,  11.  ],
       [  0.  ,   0.  ,   0.  ,   0.  ,  11.  ,  37.32]])

In [5]:
# print some properties
print(f'density = {quartz.density}')
print(f'Shear modulus (Hill average) = {quartz.G_hill:.2f}')
print(f'Pugh ratio (Hill) = {quartz.K_hill / quartz.G_hill:.2f}')

density = 2.763
Shear modulus (Hill average) = 46.57
Pugh ratio (Hill) = 1.06


In [6]:
# summarize isotropic seismic properties
quartz.wavespeeds

Unit:km/s        Vp        Vs   Vp/vs
0     Voigt  6.464861  4.207462  1.5365
1      Hill  6.356828  4.105668  1.5483
2     Reuss  6.246927  4.001284  1.5612

In [7]:
# summarize average elastic properties
quartz.elastic

Unit:GPa  Bulk_modulus  Shear_modulus  Poisson_ratio
0    Voigt     50.261111      48.912667       0.132597
1     Hill     49.551424      46.574528       0.142154
2    Reuss     48.841737      44.236389       0.152160

> **Hint**. You can save this information as text files using the method ``to_csv()``, for example: ``quartz.wavespeeds.to_csv("filename.csv")``

In [8]:
# decompose the elastic tensor into lower symmetry classes according to Browaeys and Chevrot (2004) method
quartz.decompose

{'isotropic': array([[115.48,  17.65,  17.65,   0.  ,   0.  ,   0.  ],
        [ 17.65, 115.48,  17.65,   0.  ,   0.  ,   0.  ],
        [ 17.65,  17.65, 115.48,   0.  ,   0.  ,   0.  ],
        [  0.  ,   0.  ,   0.  ,  48.91,   0.  ,   0.  ],
        [  0.  ,   0.  ,   0.  ,   0.  ,  48.91,   0.  ],
        [  0.  ,   0.  ,   0.  ,   0.  ,   0.  ,  48.91]]),
 'hexagonal': array([[-21.96,   1.23,   6.78,   0.  ,   0.  ,   0.  ],
        [  1.23, -21.96,   6.78,   0.  ,   0.  ,   0.  ],
        [  6.78,   6.78,  14.35,   0.  ,   0.  ,   0.  ],
        [  0.  ,   0.  ,   0.  ,  13.19,   0.  ,   0.  ],
        [  0.  ,   0.  ,   0.  ,   0.  ,  13.19,   0.  ],
        [  0.  ,   0.  ,   0.  ,   0.  ,   0.  , -11.59]]),
 'tetragonal': array([[-0.,  0.,  0.,  0.,  0.,  0.],
        [ 0., -0.,  0.,  0.,  0.,  0.],
        [ 0.,  0.,  0.,  0.,  0.,  0.],
        [ 0.,  0.,  0.,  0.,  0.,  0.],
        [ 0.,  0.,  0.,  0.,  0.,  0.],
        [ 0.,  0.,  0.,  0.,  0.,  0.]]),
 'orthorhombic': a

In [9]:
# show the isotropic part of the elastic tensor
quartz.decompose['isotropic']

array([[115.48,  17.65,  17.65,   0.  ,   0.  ,   0.  ],
       [ 17.65, 115.48,  17.65,   0.  ,   0.  ,   0.  ],
       [ 17.65,  17.65, 115.48,   0.  ,   0.  ,   0.  ],
       [  0.  ,   0.  ,   0.  ,  48.91,   0.  ,   0.  ],
       [  0.  ,   0.  ,   0.  ,   0.  ,  48.91,   0.  ],
       [  0.  ,   0.  ,   0.  ,   0.  ,   0.  ,  48.91]])

In [10]:
# percentages of decomposed elastic tensor
quartz.percent

{'isotropic': 72.39,
 'hexagonal': 15.62,
 'tetragonal': 0.0,
 'orthorhombic': 0.0,
 'monoclinic': 0.0,
 'others': 11.99,
 'anisotropic': 27.61}


These percentages above indicate that, according to the method of Browaeys and Chevrot (2004), the elastic tensor of quartz under these conditions can be decomposed into an isotropic part of 72.4% and an anisotropic part of 27.6%. Of the anisotropic part, 15.6% has hexagonal symmetry and 12.0% trigonal symmetry (here labelled "others"). Adding together the isotropic, hexagonal and other parts gives the undecomposed elastic tensor.

---

TODO

In [11]:
db.omphacite(1.8)

<class 'ElasticClass.ElasticProps'>

Mineral Name: Omphacite
Reference Source: https://doi.org/10.1029/2018JB016964
Crystal System: Monoclinic
Pressure (GPa): 1.8
Temperature (°C): 25
Density (g/cm3): 3.388

Elastic Tensor (Cij) in GPa:
[[245.8   92.34  81.79   0.     6.62   0.  ]
 [ 92.34 210.    64.46   0.     4.48   0.  ]
 [ 81.79  64.46 267.18   0.    36.     0.  ]
 [  0.     0.     0.    80.88   0.     6.3 ]
 [  6.62   4.48  36.     0.    70.57   0.  ]
 [  0.     0.     0.     6.3    0.    78.15]]

Tensor decomposition (Browaeys & Chevrot approach):
    Isotropy = 76.6 %
    Anisotropy = 23.4 %

Anisotropy indexes:
    Universal Elastic Anisotropy:           0.248
    Kube's Anisotropy Index (proportional): 0.099

Calculated elastic average properties:
Unit:GPa  Bulk_modulus  Shear_modulus  Poisson_ratio
   Voigt       133.351         78.213          0.255
    Hill       131.057         76.617          0.255
   Reuss       128.763         75.022          0.256

Isotropic seismic p

In [12]:
db.diopside(1.8)

<class 'ElasticClass.ElasticProps'>

Mineral Name: Diopside
Reference Source: https://doi.org/10.1016/j.pepi.2013.12.011
Crystal System: Monoclinic
Pressure (GPa): 1.8
Temperature (°C): 25
Density (g/cm3): 3.313

Elastic Tensor (Cij) in GPa:
[[240.6   86.27  77.58   0.     9.64   0.  ]
 [ 86.27 190.59  63.43   0.     3.98   0.  ]
 [ 77.58  63.43 257.9    0.    38.72   0.  ]
 [  0.     0.     0.    82.06   0.     4.09]
 [  9.64   3.98  38.72   0.    71.81   0.  ]
 [  0.     0.     0.     4.09   0.    81.56]]

Tensor decomposition (Browaeys & Chevrot approach):
    Isotropy = 74.4 %
    Anisotropy = 25.6 %

Anisotropy indexes:
    Universal Elastic Anisotropy:           0.289
    Kube's Anisotropy Index (proportional): 0.115

Calculated elastic average properties:
Unit:GPa  Bulk_modulus  Shear_modulus  Poisson_ratio
   Voigt       127.072         77.873          0.246
    Hill       124.168         76.085          0.246
   Reuss       121.265         74.297          0.246

Isotropic seis

In [13]:
db.enstatite(1.8)

<class 'ElasticClass.ElasticProps'>

Mineral Name: Enstatite
Reference Source: https://doi.org/10.1002/2016GL069963
Crystal System: Orthorhombic
Pressure (GPa): 1.8
Temperature (°C): 25
Density (g/cm3): 3.337

Elastic Tensor (Cij) in GPa:
[[252.37  90.97  74.92   0.     0.     0.  ]
 [ 90.97 192.45  67.71   0.     0.     0.  ]
 [ 74.92  67.71 244.64   0.     0.     0.  ]
 [  0.     0.     0.    86.8    0.     0.  ]
 [  0.     0.     0.     0.    81.35   0.  ]
 [  0.     0.     0.     0.     0.    84.81]]

Tensor decomposition (Browaeys & Chevrot approach):
    Isotropy = 85.2 %
    Anisotropy = 14.8 %

Anisotropy indexes:
    Universal Elastic Anisotropy:           0.106
    Kube's Anisotropy Index (proportional): 0.043

Calculated elastic average properties:
Unit:GPa  Bulk_modulus  Shear_modulus  Poisson_ratio
   Voigt       128.518         80.983           0.24
    Hill       127.550         80.263           0.24
   Reuss       126.583         79.543           0.24

Isotropic seismic

In [14]:
db.forsterite_ZB2016(P=1.8)

<class 'ElasticClass.ElasticProps'>

Mineral Name: Forsterite
Reference Source: https://doi.org/10.1002/2016GL069949
Crystal System: Orthorhombic
Pressure (GPa): 1.8
Temperature (°C): 1027
Density (g/cm3): 3.285

Elastic Tensor (Cij) in GPa:
[[282.82  72.83  67.77   0.     0.     0.  ]
 [ 72.83 183.61  71.08   0.     0.     0.  ]
 [ 67.77  71.08 208.99   0.     0.     0.  ]
 [  0.     0.     0.    57.3    0.     0.  ]
 [  0.     0.     0.     0.    69.17   0.  ]
 [  0.     0.     0.     0.     0.    68.96]]

Tensor decomposition (Browaeys & Chevrot approach):
    Isotropy = 80.2 %
    Anisotropy = 19.8 %

Anisotropy indexes:
    Universal Elastic Anisotropy:           0.187
    Kube's Anisotropy Index (proportional): 0.075

Calculated elastic average properties:
Unit:GPa  Bulk_modulus  Shear_modulus  Poisson_ratio
   Voigt       122.087         70.002          0.259
    Hill       120.458         68.922          0.260
   Reuss       118.829         67.842          0.260

Isotropic seis

In [15]:
db.forsterite_Mao(P=1.8)

<class 'ElasticClass.ElasticProps'>

Mineral Name: Forsterite
Reference Source: https://doi.org/10.1016/j.epsl.2015.06.045
Crystal System: Orthorhombic
Pressure (GPa): 1.8
Temperature (°C): 627
Density (g/cm3): 3.301

Elastic Tensor (Cij) in GPa:
[[310.33  53.85  67.18   0.     0.     0.  ]
 [ 53.85 189.12  78.3    0.     0.     0.  ]
 [ 67.18  78.3  220.88   0.     0.     0.  ]
 [  0.     0.     0.    60.23   0.     0.  ]
 [  0.     0.     0.     0.    74.11   0.  ]
 [  0.     0.     0.     0.     0.    73.78]]

Tensor decomposition (Browaeys & Chevrot approach):
    Isotropy = 75.9 %
    Anisotropy = 24.1 %

Anisotropy indexes:
    Universal Elastic Anisotropy:           0.284
    Kube's Anisotropy Index (proportional): 0.115

Calculated elastic average properties:
Unit:GPa  Bulk_modulus  Shear_modulus  Poisson_ratio
   Voigt       124.332         76.357          0.245
    Hill       122.586         74.507          0.247
   Reuss       120.841         72.656          0.250

Isotropic

In [16]:
db.zoisite()

<class 'ElasticClass.ElasticProps'>

Mineral Name: Zoisite
Reference Source: https://doi.org/10.2138/am.2007.2329
Crystal System: Orthorhombic
Pressure (GPa): 0.0
Temperature (°C): 25
Density (g/cm3): 3.343

Elastic Tensor (Cij) in GPa:
[[279.8  94.7  88.7   0.    0.    0. ]
 [ 94.7 249.2  27.5   0.    0.    0. ]
 [ 88.7  27.5 209.4   0.    0.    0. ]
 [  0.    0.    0.   51.8   0.    0. ]
 [  0.    0.    0.    0.   81.4   0. ]
 [  0.    0.    0.    0.    0.   66.3]]

Tensor decomposition (Browaeys & Chevrot approach):
    Isotropy = 74.4 %
    Anisotropy = 25.6 %

Anisotropy indexes:
    Universal Elastic Anisotropy:           0.373
    Kube's Anisotropy Index (proportional): 0.148

Calculated elastic average properties:
Unit:GPa  Bulk_modulus  Shear_modulus  Poisson_ratio
   Voigt       128.911         75.067          0.256
    Hill       125.274         72.855          0.256
   Reuss       121.637         70.643          0.257

Isotropic seismic properties:
Unit:km/s    Vp    Vs  Vp

In [17]:
db.chlorite(1.8)

<class 'ElasticClass.ElasticProps'>

Mineral Name: Chlorite
Reference Source: https://doi.org/10.1002/2014GL059334
Crystal System: Monoclinic
Pressure (GPa): 1.8
Temperature (°C): 25
Density (g/cm3): 2.594

Elastic Tensor (Cij) in GPa:
[[199.1   57.76  23.51   0.     3.55   0.  ]
 [ 57.76 201.96  31.01   0.     0.05   0.  ]
 [ 23.51  31.01 162.46   0.     0.78   0.  ]
 [  0.     0.     0.    24.55   0.     0.33]
 [  3.55   0.05   0.78   0.    23.81   0.  ]
 [  0.     0.     0.     0.33   0.    71.77]]

Tensor decomposition (Browaeys & Chevrot approach):
    Isotropy = 72.6 %
    Anisotropy = 27.4 %

Anisotropy indexes:
    Universal Elastic Anisotropy:           1.728
    Kube's Anisotropy Index (proportional): 0.655

Calculated elastic average properties:
Unit:GPa  Bulk_modulus  Shear_modulus  Poisson_ratio
   Voigt        87.564         54.109          0.244
    Hill        86.392         47.242          0.269
   Reuss        85.220         40.376          0.295

Isotropic seismic pr

In [18]:
db.pyrope(1.8)

<class 'ElasticClass.ElasticProps'>

Mineral Name: Pyrope
Reference Source: https://doi.org/10.1016/j.epsl.2012.11.041
Crystal System: Cubic
Pressure (GPa): 1.8
Temperature (°C): 750
Density (g/cm3): 3.740

Elastic Tensor (Cij) in GPa:
[[284.1  107.2  107.2    0.     0.     0.  ]
 [107.2  284.1  107.2    0.     0.     0.  ]
 [107.2  107.2  284.1    0.     0.     0.  ]
 [  0.     0.     0.    90.26   0.     0.  ]
 [  0.     0.     0.     0.    90.26   0.  ]
 [  0.     0.     0.     0.     0.    90.26]]

Tensor decomposition (Browaeys & Chevrot approach):
    Isotropy = 99.1 %
    Anisotropy = 0.9 %

Anisotropy indexes:
    Universal Elastic Anisotropy:           0.000
    Kube's Anisotropy Index (proportional): 0.000

Calculated elastic average properties:
Unit:GPa  Bulk_modulus  Shear_modulus  Poisson_ratio
   Voigt       166.167         89.536          0.272
    Hill       166.167         89.532          0.272
   Reuss       166.167         89.527          0.272

Isotropic seismic pro

In [19]:
db.plagioclase('An37')

<class 'ElasticClass.ElasticProps'>

Mineral Name: plagioclase_An37
Reference Source: https://doi.org/10.1002/2015JB012736
Crystal System: Triclinic
Pressure (GPa): 0.0
Temperature (°C): 25
Density (g/cm3): 2.666

Elastic Tensor (Cij) in GPa:
[[ 96.2  46.1  38.4   5.9  -0.2  -0.4]
 [ 46.1 189.4  15.4  -7.   -5.1  -6.8]
 [ 38.4  15.4 171.9   2.2   7.2  -9.8]
 [  5.9  -7.    2.2  23.6  -1.1  -4.8]
 [ -0.2  -5.1   7.2  -1.1  33.1   1.4]
 [ -0.4  -6.8  -9.8  -4.8   1.4  35.5]]

Tensor decomposition (Browaeys & Chevrot approach):
    Isotropy = 57.7 %
    Anisotropy = 42.3 %

Anisotropy indexes:
    Universal Elastic Anisotropy:           1.364
    Kube's Anisotropy Index (proportional): 0.521

Calculated elastic average properties:
Unit:GPa  Bulk_modulus  Shear_modulus  Poisson_ratio
   Voigt        73.033         42.280          0.257
    Hill        70.920         37.910          0.273
   Reuss        68.807         33.539          0.290

Isotropic seismic properties:
Unit:km/s    Vp    

In [20]:
db.amphiboles('Pargasite')

<class 'ElasticClass.ElasticProps'>

Mineral Name: Pargasite
Reference Source: https://doi.org/10.1016/j.pepi.2016.10.010
Crystal System: Monoclinic
Pressure (GPa): 0.0
Temperature (°C): 25
Density (g/cm3): 3.163

Elastic Tensor (Cij) in GPa:
[[141.6  57.1  49.6   0.   -0.2   0. ]
 [ 57.1 197.8  60.9   0.  -10.9   0. ]
 [ 49.6  60.9 225.4   0.  -31.4   0. ]
 [  0.    0.    0.   75.8   0.    3.3]
 [ -0.2 -10.9 -31.4   0.   49.9   0. ]
 [  0.    0.    0.    3.3   0.   51.7]]

Tensor decomposition (Browaeys & Chevrot approach):
    Isotropy = 69.4 %
    Anisotropy = 30.6 %

Anisotropy indexes:
    Universal Elastic Anisotropy:           0.493
    Kube's Anisotropy Index (proportional): 0.194

Calculated elastic average properties:
Unit:GPa  Bulk_modulus  Shear_modulus  Poisson_ratio
   Voigt       100.000         61.960          0.243
    Hill        96.768         59.536          0.245
   Reuss        93.537         57.112          0.246

Isotropic seismic properties:
Unit:km/s    Vp    

In [21]:
db.amphiboles?

Signature: db.amphiboles(type='Hornblende')
Docstring:
Returns the corresponding elastic tensor (GPa) and density
(g/cm3) and other derived elastic properties of different
types of Ca-Na amphibole based on experimental data of
Brown and Abramson (2016) [1]

Caveats
-------
    - The function does not account for temperature or
    pressure effects and assumes room conditions

Parameters
----------
type : str
    the type of Ca-Na amphibole

Returns
-------
properties : ElasticProps dataclass
    An object containing the following properties:
    - name: Name of the crystal ('alpha_quartz').
    - crystal_system: Crystal system.
    - temperature: Temperature in degrees Celsius (assumed 25).
    - pressure: Pressure in GPa.
    - density: Density in g/cm3.
    - cijs: 6x6 array representing the elastic tensor.
    - sijs: 6x6 array representing the compliance tensor
    - reference: Reference to the source publication.
    - decompose: the decomposition of the elastic tensor
        int

In [22]:
db.antigorite()

<class 'ElasticClass.ElasticProps'>

Mineral Name: Antigorite
Reference Source: https://doi.org/10.1029/2022GL099411
Crystal System: Monoclinic
Pressure (GPa): 0.0
Temperature (°C): 25
Density (g/cm3): 2.615

Elastic Tensor (Cij) in GPa:
[[190.6  61.3  21.3   0.    2.9   0. ]
 [ 61.3 208.2  19.    0.   -1.2   0. ]
 [ 21.3  19.   85.4   0.    0.4   0. ]
 [  0.    0.    0.   13.5   0.   -3.2]
 [  2.9  -1.2   0.4   0.   20.    0. ]
 [  0.    0.    0.   -3.2   0.   67.5]]

Tensor decomposition (Browaeys & Chevrot approach):
    Isotropy = 63.4 %
    Anisotropy = 36.6 %

Anisotropy indexes:
    Universal Elastic Anisotropy:           3.296
    Kube's Anisotropy Index (proportional): 1.087

Calculated elastic average properties:
Unit:GPa  Bulk_modulus  Shear_modulus  Poisson_ratio
   Voigt        76.378         45.707          0.251
    Hill        68.741         37.055          0.272
   Reuss        61.104         28.404          0.299

Isotropic seismic properties:
Unit:km/s    Vp    Vs  V

In [23]:
import sys
from datetime import date    
today = date.today().isoformat()

print(f'Notebook tested in {today} using:')
print('Python', sys.version)
print('Numpy', np.__version__)

Notebook tested in 2023-09-06 using:
Python 3.10.12 | packaged by Anaconda, Inc. | (main, Jul  5 2023, 19:01:18) [MSC v.1916 64 bit (AMD64)]
Numpy 1.25.2
